packages/cli/src/deploy/getFunctions.ts

```diff
@@ -15,44 +13,52 @@ export async function getFunctions({
     strict: true,
     fromBlock: worldDeploy.deployBlock,
     toBlock: worldDeploy.stateBlock,
     address: worldDeploy.address,
     event: parseAbiItem(storeSetRecordEvent),
-    args: { tableId: worldTables.world_FunctionSignatures.tableId },
+    args: { tableId: worldTables.world_FunctionSelectors.tableId },
+  });
+
+  const selectors = selectorLogs.map((log) => {
+    return {
+      ...decodeValueArgs(worldTables.world_FunctionSelectors.valueSchema, log.args),
+      ...decodeKey(worldTables.world_FunctionSelectors.keySchema, log.args.keyTuple),
+    };
   });
+  debug("found", selectors.length, "function selectors for", worldDeploy.address);
 
-  const signatures = logs.map((log) => {
-    const value = decodeValueArgs(worldTables.world_FunctionSignatures.valueSchema, log.args);
-    return value.functionSignature;
+  // This assumes we only use `FunctionSignatures._set(...)`, which is true as of this writing.
+  debug("looking up function signatures for", worldDeploy.address);
+  const signatureLogs = await getLogs(client, {
+    strict: true,
+    fromBlock: worldDeploy.deployBlock,
+    toBlock: worldDeploy.stateBlock,
+    address: worldDeploy.address,
+    event: parseAbiItem(storeSetRecordEvent),
+    args: { tableId: worldTables.world_FunctionSignatures.tableId },
   });
-  debug("found", signatures.length, "function signatures for", worldDeploy.address);
 
-  // TODO: parallelize with a bulk getRecords
-  const functions = await Promise.all(
-    signatures.map(async (signature) => {
-      const selector = toFunctionSelector(signature);
-      const { systemId, systemFunctionSelector } = await getTableValue({
-        client,
-        worldDeploy,
-        table: worldTables.world_FunctionSelectors,
-        key: { worldFunctionSelector: selector },
-      });
-      const { namespace, name } = hexToResource(systemId);
-      // TODO: find away around undoing contract logic (https://github.com/latticexyz/mud/issues/1708)
-      const systemFunctionSignature = namespace === "" ? signature : signature.replace(`${namespace}_${name}_`, "");
-      return {
-        signature,
-        selector,
-        systemId,
-        systemFunctionSignature,
-        systemFunctionSelector,
-      };
+  const selectorToSignature = Object.fromEntries(
+    signatureLogs.map((log) => {
+      return [
+        decodeKey(worldTables.world_FunctionSignatures.keySchema, log.args.keyTuple).functionSelector,
+        decodeValueArgs(worldTables.world_FunctionSignatures.valueSchema, log.args).functionSignature,
+      ];
     }),
   );
+  debug("found", signatureLogs.length, "function signatures for", worldDeploy.address);
+
+  const functions = selectors.map(({ worldFunctionSelector, systemFunctionSelector, systemId }) => ({
+    selector: worldFunctionSelector,
+    signature: selectorToSignature[worldFunctionSelector],
+    systemFunctionSelector,
+    systemFunctionSignature: selectorToSignature[systemFunctionSelector],
```

Comment:
This should always exist since the function signature table includes both world and system signatures. We could throw but it would be more like a "if this throws there is a bug in the protocol".

Reply:
seems reasonable to throw when we suspect a protocol issue than letting this failed with some obscure `undefined` error